File: bob-client/app/src/main/java/fr/dmconcept/bob/client/BobApplication.java
package fr.dmconcept.bob.client;

import android.app.Application;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import java.util.ArrayList;

import fr.dmconcept.bob.client.models.BoardConfig;
import fr.dmconcept.bob.client.models.Project;
import fr.dmconcept.bob.client.models.ServoConfig;
import fr.dmconcept.bob.client.models.Step;
import fr.dmconcept.bob.client.models.dao.BoardConfigDao;
import fr.dmconcept.bob.client.models.dao.ProjectDao;
import fr.dmconcept.bob.client.models.helpers.BobSqliteOpenHelper;

public class BobApplication extends Application {

    private static final String TAG = "BobApplication";

    public static final String PREFERENCES_SERVER_IP = "sever_ip";

    BobSqliteOpenHelper mOpenHelper;
    SQLiteDatabase      mDatabase;

    BoardConfigDao mBoardConfigDao;
    ProjectDao mProjectsDao;

    @Override
    public void onCreate() {

        Log.i(TAG, "onCreate()");

        mOpenHelper = new BobSqliteOpenHelper(this);
        mDatabase   = mOpenHelper.getWritableDatabase();

        mBoardConfigDao = new BoardConfigDao(mDatabase);
        mProjectsDao    = new ProjectDao(mDatabase, mBoardConfigDao);

        if (BuildConfig.DEBUG && mProjectsDao.findAll().size() == 0)
            // First run, create the fixtures
            createFixtures();

        super.onCreate();

    }

    private void createFixtures() {

        Log.i(TAG, "DEBUG MODE - Creating servos config fixtures");

        ArrayList<ServoConfig> servoConfigs1 = new ArrayList<ServoConfig>() {{
            add(new ServoConfig(3, 558, 2472, 50));
            add(new ServoConfig(4, 558, 2472, 50));
        }};

        ArrayList<ServoConfig> servoConfigs2 = new ArrayList<ServoConfig>() {{
            add(new ServoConfig(3, 558, 2472, 50));
            add(new ServoConfig(4, 558, 2472, 50));
            add(new ServoConfig(5, 558, 2472, 50));
            add(new ServoConfig(6, 558, 2472, 50));
            add(new ServoConfig(7, 558, 2472, 50));
            add(new ServoConfig(8, 558, 2472, 50));
            add(new ServoConfig(9, 558, 2472, 50));
        }};

        BoardConfig config1 = new BoardConfig("Servos on pins 3 and 4"  , servoConfigs1);
        BoardConfig config2 = new BoardConfig("Servos on pins 3 to 9", servoConfigs2);

        config1.setId(mBoardConfigDao.save(config1));
        config2.setId(mBoardConfigDao.save(config2));

        Log.i(TAG, "DEBUG MODE - Creating project...");

        // Project 1
        mProjectsDao.create(
            new Project(-1, "Simple demo project", config1, new ArrayList<Step>() {{
                add(new Step(4000, new ArrayList<Integer>() {{ add( 1); add(50); }}));
                add(new Step(2000, new ArrayList<Integer>() {{ add(99); add(50); }}));
                add(new Step(4000, new ArrayList<Integer>() {{ add(50); add(50); }}));
                add(new Step(0   , new ArrayList<Integer>() {{ add( 1); add(50); }}));
            }})
        );

        // Project 2
        mProjectsDao.create(
            new Project(-1, "Bob demo project", config2, new ArrayList<Step>() {{
                add(new Step(5000, new ArrayList<Integer>() {{
                    add(  0); add(20); add(20); add(10); add(20); add(80); add( 100);
                }}));
                add(new Step(6000, new ArrayList<Integer>() {{
                    add( 25); add(40); add(40); add(20); add(40); add(60); add( 50);
                }}));
                add(new Step(5000, new ArrayList<Integer>() {{
                    add( 75); add(60); add(60); add(30); add(60); add(40); add( 25);
                }}));
                add(new Step(10000, new ArrayList<Integer>() {{
                    add(100); add(80); add(80); add(40); add(80); add(20); add(  0);
                }}));
            }})
        );

        Log.i(TAG, "DEBUG MODE - Fixtures creation done");

    }

    public ProjectDao getProjectsDao() {
        return mProjectsDao;
    }

    public BoardConfigDao getBoardConfigDao() {
        return mBoardConfigDao;
    }

    @Override
    public void onTerminate() {

        // Close the database and the DB helper
        mDatabase.close();
        mOpenHelper.close();

        super.onTerminate();

    }


}

File: bob-client/src/main/scala/fr/dmconcept/bob/client/BobApplication.scala
package fr.dmconcept.bob.client

import android.app.Application
import android.database.sqlite.SQLiteDatabase
import fr.dmconcept.bob.client.models.dao.{ProjectDao, BoardConfigDao}
import fr.dmconcept.bob.client.models.helpers.BobSqliteOpenHelper
import fr.dmconcept.bob.client.models.{Step, Project, ServoConfig, BoardConfig}
import java.util.UUID
import org.scaloid.common._

object BobApplication {

  object Preferences {
    val SERVER_IP = "serverIP"
    val AUTOPLAY  = "autoplay"
  }

  object Params {
    val PORT = 8000
  }

}

class BobApplication extends Application with TagUtil {

  implicit override val loggerTag = LoggerTag("BobClient")

  lazy val sqliteOpenHelper : BobSqliteOpenHelper = new BobSqliteOpenHelper(this)
  lazy val sqliteDatabase   : SQLiteDatabase      = sqliteOpenHelper.getWritableDatabase

  lazy val boardConfigDao   : BoardConfigDao      = new BoardConfigDao(sqliteDatabase)
  lazy val projectsDao      : ProjectDao          = new ProjectDao    (sqliteDatabase, boardConfigDao)

  override def onCreate() {

    info("BobApplication.onCreate()")

    if (/*BuildConfig.DEBUG && */ projectsDao.findAll().isEmpty)
      firstRun()

    super.onCreate()

  }

  private def firstRun() {

    info("BobApplication.onCreate() First run, creating the servos config fixtures")

    def servoConfigs(servoPins: List[Int]): Vector[ServoConfig] = servoPins.map { s =>
      ServoConfig(s"Servo $s", s)
    }.toVector

    val configs = List(
      BoardConfig("Servos on pins 3 and 4", servoConfigs(List(3, 4)                 )),
      BoardConfig("Servos on pins 3 to 9" , servoConfigs(List(3, 4, 5, 6, 7, 10, 11)))
    )

    configs.foreach(boardConfigDao.create)

    info("BobApplication.onCreate() First run, creating the project fixtures")

    // Project 1
    projectsDao.create(
      Project(UUID.randomUUID.toString, "Simple demo project", configs(0), Vector(
        Step(4000, Vector( 1, 50)),
        Step(2000, Vector(99, 50)),
        Step(4000, Vector(50, 50)),
        Step(0   , Vector( 1, 50))
      ))
    )

    // Project 2
    projectsDao.create(
      Project(UUID.randomUUID.toString, "Bob demo project", configs(1), Vector(
        Step( 500, Vector(  0, 100,   0, 100,   0,  40,   0)),
        Step(1000, Vector(100,   0,  20,  80, 100,  60,  20)),
        Step( 200, Vector(  0, 100,  40,  60,  20,  20,  40)),
        Step( 500, Vector(100,   0,  60,  40,  80, 800,  60)),
        Step( 200, Vector(  0, 100,  80,  20,  40,   0,  80)),
        Step(  0 , Vector(100,   0, 100,   0,  60, 100, 100)),

        Step( 500, Vector(  0, 100,   0, 100,   0,  40,   0)),
        Step(1000, Vector(100,   0,  20,  80, 100,  60,  20)),
        Step( 200, Vector(  0, 100,  40,  60,  20,  20,  40)),
        Step( 500, Vector(100,   0,  60,  40,  80, 800,  60)),
        Step( 200, Vector(  0, 100,  80,  20,  40,   0,  80)),
        Step(  0 , Vector(100,   0, 100,   0,  60, 100, 100))
      ))
    )

  }

  override def onTerminate() {

    info("BobApplication.onTerminate() Closing DB and open helper")

    // Close the database and the DB helper
    sqliteDatabase.close()
    sqliteOpenHelper.close()

    super.onTerminate()

  }

}
